### PR TITLE
Add bundle prepare

### PR DESCRIPTION
**Pull request by djwhatle**
- Encodes `apb.yml` in current directory as b64
 - Writes b64 contents into existing label on Dockerfile
 - Attempts to write same format as legacy python `apb` tool (produces no changes in git for APB examples I've tried)

### CODE REVIEW

**Comment by djwhatle** [comment]
Addresses https://github.com/automationbroker/sbcli/pull/62

File: cmd/bundle.go [hunk 1]
```diff
@@ -239,3 +261,95 @@ func showBundleInfo(bundleName string, registryName string) {
...
+	bundleMetaSection := containerMeta[lineStartIndex : blobStartIndex-1]
+	if !bytes.Contains(bundleMetaSection, []byte(lineBreakText)) {
+		bundleMetaSection = append(bundleMetaSection, lineBreakText...)
+	}
+	bundleMetaSection = append(bundleMetaSection, byte('"'))
+	bundleMetaSection = append(bundleMetaSection, bundleMetaEncoded...)
+
+	// Build final output with linebreaked "LABEL" b64 content
+	newContainerMeta := containerMeta[0:lineStartIndex]
+	newContainerMeta = append(newContainerMeta, []byte(bundleMetaSection)...)
```
Comment (djwhatle) [inline cmd/bundle.go]: I don't think I need this `[]byte` cast anymore. Leftover from when I was using strings.

File: cmd/bundle.go [hunk 2]
```diff
@@ -239,3 +261,95 @@ func showBundleInfo(bundleName string, registryName string) {
...
+	blobEndOffset := bytes.IndexByte(containerMeta[blobStartIndex:], byte('"'))
+	if blobEndOffset == -1 {
+		log.Errorf("Didn't find end of bundle label in container metadata file")
+		return []byte{}
+	}
+	blobEndIndex := blobStartIndex + blobEndOffset
+
+	// Build new label section
+	bundleMetaSection := containerMeta[lineStartIndex : blobStartIndex-1]
+	if !bytes.Contains(bundleMetaSection, []byte(lineBreakText)) {
```
Comment (djwhatle) [inline cmd/bundle.go]: Also don't think I need this one

File: cmd/bundle.go [hunk 3]
```diff
@@ -239,3 +261,95 @@ func showBundleInfo(bundleName string, registryName string) {
...
+	if errB != nil {
+		log.Errorf("Bundle metadata file [%s] not found in working directory", bundleMetaFilename)
+	}
+	containerMeta, errC := ioutil.ReadFile(containerMetaPath)
+	if errC != nil {
+		log.Errorf("Container metadata file [%s] not found in working directory", containerMetaFilename)
+	}
+	if errB != nil || errC != nil {
+		return
+	}
```
[the remark this answers is not in the document]
Comment (jmrodri) [reply cmd/bundle.go]: What you have is correct, just stylistic feels weird with the errB.

File: cmd/bundle.go [hunk 4]
```diff
@@ -85,6 +103,10 @@ var bundleDeprovisionCmd = &cobra.Command{
 func init() {
 	rootCmd.AddCommand(bundleCmd)
 
+	bundlePrepareCmd.Flags().StringVarP(&bundleMetadataFilename, "bmeta", "b", "apb.yml", "Bundle metadata file to encode as b64")
+	bundlePrepareCmd.Flags().StringVarP(&containerMetadataFilename, "cmeta", "c", "Dockerfile", "Container metadata file to stamp")
```
[the remark this answers is not in the document]
Comment (djwhatle) [reply cmd/bundle.go]: Good point, will change this.

File: cmd/bundle.go [hunk 5]
```diff
@@ -239,3 +261,95 @@ func showBundleInfo(bundleName string, registryName string) {
 
 	return
 }
+
+// stampBundleMetadata encodes bundle metadata (e.g. apb.yml) as b64 and stamps onto container build file (e.g. Dockerfile)
```
[the remark this answers is not in the document]
Comment (djwhatle) [reply cmd/bundle.go]: ACK, will fix this

File: cmd/bundle.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (djwhatle) [reply cmd/bundle.go]: My idea to not require the user to run `sbcli` multiple time to see all errors. I see your point though, and the gain from the user side is kind of small in this case. I'll rethink this.

**Comment by djwhatle** [comment]
@jmrodri pushed updates in response to comments

**Comment by dymurray** [comment]
```
[dymurray@dymurray foo2]$ sbcli bundle prepare
ERRO Didn't find expected bundle label in container metadata file 
ERRO Failed to add [apb.yml] contents to [Dockerfile] 
```

**Comment by dymurray** [comment]
```
[dymurray@dymurray foo2]$ cat Dockerfile 
FROM ansibleplaybookbundle/apb-base

LABEL "com.redhat.bundle.spec"=\

COPY playbooks /opt/apb/actions
COPY roles /opt/ansible/roles
RUN chmod -R g=u /opt/{ansible,apb}
USER apb
```

**Comment by dymurray** [comment]
```
[dymurray@dymurray constellations]$ ansible-galaxy init --type apb foo4-apb                              
- foo4-apb was created successfully
[dymurray@dymurray constellations]$ cd foo4-apb/
[dymurray@dymurray foo4-apb]$ ls
apb.yml   Dockerfile  handlers  meta       README.md  templates  vars
defaults  files       Makefile  playbooks  tasks      tests
[dymurray@dymurray foo4-apb]$ cat Dockerfile 
FROM ansibleplaybookbundle/apb-base

LABEL "com.redhat.apb.spec"=\

COPY playbooks /opt/apb/actions
COPY . /opt/ansible/roles/foo4-apb
RUN chmod -R g=u /opt/{ansible,apb}
USER apb[dymurray@dymurray foo4-apb]$ sbcli bundle prepare
ERRO Didn't find expected bundle label in container metadata file 
ERRO Failed to add [apb.yml] contents to [Dockerfile] 
```

**Comment by djwhatle** [comment]
@dymurray @jmrodri it's worth noting that the `Dockerfile`s generated by `apb init` and `ansible-galaxy init --type apb` are invalid. They have a malformed `LABEL` line for the b64 apb spec, and trying to run `docker build` without making modifications will fail. This is why the regex currently fails in these scenarios.

That being said, it still makes sense to me to support the format that currently comes out of these tools for backwards-compatibility sake, but I am thinking of submitting PRs to make the initial Dockerfile from `apb` and `galaxy` valid.

**Comment by djwhatle** [comment]
@dymurray added a handler for the format provided by `apb init` and `galaxy init`

**Comment by djwhatle** [comment]
I'm going to look at reworking this a little bit to use a slice `insert` instead of building new slices when I want to make changes.

**Comment by dymurray** [comment]
```
[dymurray@dymurray minipos-apb]$ sbcli bundle prepare
Wrote b64 encoded [apb.yml] to [Dockerfile]
```

**Pull request merged**